File: game_plugin/src/Logger.cpp
#include "sr2ap/Logger.hpp"

#include "rust/cxx.h"
#include "sr2ap/src/ffi/logger_ffi.rs.h"

#include <cstdint>
#include <string_view>

namespace sr2ap::log {
    namespace {
        [[nodiscard]] constexpr rust::LogLevel ToRustLevel(
            const Level level) noexcept {
            switch (level) {
                case Level::Trace:
                    return rust::LogLevel::Trace;
                case Level::Debug:
                    return rust::LogLevel::Debug;
                case Level::Info:
                    return rust::LogLevel::Info;
                case Level::Warning:
                    return rust::LogLevel::Warning;
                case Level::Error:
                    return rust::LogLevel::Error;
                case Level::Critical:
                    return rust::LogLevel::Critical;
            }

            return rust::LogLevel::Error;
        }

        [[nodiscard]] ::rust::Slice<const std::uint8_t> Bytes(
            const std::string_view value) noexcept {
            return {
                reinterpret_cast<const std::uint8_t*>(value.data()),
                value.size(),
            };
        }
    }

    bool Initialize(
        const std::filesystem::path& directory,
        const bool debugEnabled) {
        const auto& native = directory.native();

        try {
            rust::log_initialize(
                {reinterpret_cast<const std::uint16_t*>(native.data()),
                 native.size()},
                debugEnabled);
            return true;
        } catch (const ::rust::Error&) {
            return false;
        }
    }

    void Write(
        const Level level,
        const std::string_view subsystem,
        const std::string_view message) {
        try {
            rust::log_write(
                ToRustLevel(level),
                Bytes(subsystem),
                Bytes(message));
        } catch (const ::rust::Error&) {
            // Logging cannot report a logging failure through the logger itself.
        }
    }

    void Flush() {
        try {
            rust::log_flush();
        } catch (const ::rust::Error&) {
        }
    }

    void Shutdown() {
        rust::log_shutdown();
    }
}

File: game_plugin/rust/src/logger.rs
use std::{fmt, io::Write, path::Path, sync::OnceLock};

use flexi_logger::{DeferredNow, FileSpec, Logger, LoggerHandle};
use log::{Level, Record};

#[derive(Debug, thiserror::Error)]
pub(crate) enum LoggerError {
    #[error("logger is not initialized")]
    NotInitialized,
    #[error("could not install the global Rust logger")]
    Install,
    #[error(transparent)]
    Backend(#[from] flexi_logger::FlexiLoggerError),
    #[error("Invalid argument")]
    InvalidArgument,
}

static LOGGER: OnceLock<LoggerHandle> = OnceLock::new();

pub(crate) fn initialize(directory: &Path, debug_enabled: bool) -> Result<(), LoggerError> {
    let specification = if debug_enabled { "trace" } else { "info" };

    let handle = Logger::try_with_str(specification)?
        .log_to_file(FileSpec::try_from(directory.join("SR2Archipelago.log"))?)
        .format_for_files(format_record)
        .start()?;

    LOGGER.set(handle).map_err(|_| LoggerError::Install)
}

pub(crate) fn write(
    level: Level,
    subsystem: &str,
    message: impl fmt::Display,
) -> Result<(), LoggerError> {
    LOGGER.get().ok_or(LoggerError::NotInitialized)?;
    log::log!(target: subsystem, level, "{message}");
    Ok(())
}

pub(crate) fn flush() -> Result<(), LoggerError> {
    LOGGER.get().ok_or(LoggerError::NotInitialized)?.flush();
    Ok(())
}

pub(crate) fn shutdown() {
    if let Some(logger) = LOGGER.get() {
        logger.shutdown();
    }
}

fn level_name(level: Level) -> &'static str {
    match level {
        Level::Trace => "trace",
        Level::Debug => "debug",
        Level::Info => "info",
        Level::Warn => "warning",
        Level::Error => "error",
    }
}

fn format_record(
    writer: &mut dyn Write,
    now: &mut DeferredNow,
    record: &Record<'_>,
) -> std::io::Result<()> {
    writeln!(
        writer,
        "[{}] [{}] [{}] {}",
        now.format("%H:%M:%S%.3f"),
        level_name(record.level()),
        record.target(),
        record.args()
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn format_record_should_include_level_target_and_message() {
        let mut output = Vec::new();
        let mut now = DeferredNow::new();
        let record = Record::builder()
            .level(Level::Warn)
            .target("Test")
            .args(format_args!("record"))
            .build();

        format_record(&mut output, &mut now, &record).expect("format record");
        let output = String::from_utf8(output).expect("valid UTF-8 log record");

        assert!(output.contains("[warning] [Test] record"));
    }
}
